### PR TITLE
Fixed #29138 -- Added ModelAdmin.autocomplete_fields support for ForeignKeys that use to_field.

### PR DESCRIPTION
**Pull request by ConstantinoSchillebeeckx**
Bug fix for [29138](https://code.djangoproject.com/ticket/29138); creating [PR as requested](https://code.djangoproject.com/ticket/29138#comment:10).

> **Note** I've been able to update all unit tests and create some new ones, however I haven't been able to leverage `to_field_allowed` to prevent data leaks. I've tried to implement it (see ​here), however I can't get it to play nicely with the unit tests. When uncommented, the `id_field` isn't properly being considered as to_field_allowed. I'm not familiar with this function, so could use some help troubleshooting.

### CODE REVIEW

File: django/contrib/admin/views/autocomplete.py [hunk 1]
```diff
@@ -24,12 +24,15 @@ def get(self, request, *args, **kwargs):
             return JsonResponse({'error': '403 Forbidden'}, status=403)
 
         self.term = request.GET.get('term', '')
+        self.id_field = request.GET.get('id_field', 'pk')
+        # set default id_field to 'pk' if not provided or not to_field_allowed()
+        #self.id_field = self.id_field if self.model_admin.to_field_allowed(request, self.id_field) else 'pk' TODO: not working
         self.paginator_class = self.model_admin.paginator
-        self.object_list = self.get_queryset()
+        self.object_list = self.get_queryset().order_by(self.id_field)
```
Comment (charettes) [inline django/contrib/admin/views/autocomplete.py]: I don't see a reason for the `order_by` here.

File: django/contrib/admin/views/autocomplete.py [hunk 2]
```diff
@@ -24,12 +24,15 @@ def get(self, request, *args, **kwargs):
             return JsonResponse({'error': '403 Forbidden'}, status=403)
 
         self.term = request.GET.get('term', '')
+        self.id_field = request.GET.get('id_field', 'pk')
+        # set default id_field to 'pk' if not provided or not to_field_allowed()
+        #self.id_field = self.id_field if self.model_admin.to_field_allowed(request, self.id_field) else 'pk' TODO: not working
```
Comment (charettes) [inline django/contrib/admin/views/autocomplete.py]: Stick to the existing pattern to validate `to_field`

```python
to_field = request.GET.get(TO_FIELD_VAR)
if to_field:
    if not self.model_admin.to_field_allowed(request, to_field):
        raise DisallowedModelAdminToField("The field %s cannot be referenced." % to_field)
else:
    to_field = 'pk'
```

See https://github.com/django/django/blob/0cd465b63aa7c03a3d14bd5fd6543628d585f8da/django/contrib/admin/options.py#L1529-L1531

File: django/contrib/admin/widgets.py [hunk 3]
```diff
@@ -385,10 +385,13 @@ def __init__(self, rel, admin_site, attrs=None, choices=(), using=None):
         self.db = using
         self.choices = choices
         self.attrs = {} if attrs is None else attrs.copy()
+        self.id_field = self.rel.field.to_fields[0] if hasattr(self.rel.field, 'to_fields') else 'pk'
```
Comment (charettes) [inline django/contrib/admin/widgets.py]: Use `to_field`.

File: django/contrib/admin/widgets.py [hunk 4]
```diff
@@ -385,10 +385,13 @@ def __init__(self, rel, admin_site, attrs=None, choices=(), using=None):
...
+        
 
     def get_url(self):
         model = self.rel.model
-        return reverse(self.url_name % (self.admin_site.name, model._meta.app_label, model._meta.model_name))
+        url = reverse(self.url_name % (self.admin_site.name, model._meta.app_label, model._meta.model_name))
+        return url + '?id_field=' + self.id_field if self.id_field else url
```
Comment (charettes) [inline django/contrib/admin/widgets.py]: From what I can see `self.id_field` will never be `False` from what I can see?

File: django/contrib/admin/views/autocomplete.py (hunk 1 above)
Comment (ConstantinoSchillebeeckx) [reply django/contrib/admin/views/autocomplete.py]: I was getting warnings about unordered querysets and thought, to return consistent query results for a given term, to order the results.

File: django/contrib/admin/views/autocomplete.py (hunk 2 above)
Comment (ConstantinoSchillebeeckx) [reply django/contrib/admin/views/autocomplete.py]: I implemented things as you suggested however I can't get the unit test `test_autocomplete_view` to pass since it raises this error. 

I don't understand why the field `fk` on the model [`ParentWithFK`](https://github.com/django/django/blob/master/tests/admin_views/models.py#L891) is diallowed.

**Comment by charettes** [comment]
Oh by the way, you'll should base your branch off `master` and not `stable/2.1.x` and will backport will take place if necessary.

**Comment by ConstantinoSchillebeeckx** [comment]
The unit tests `admin_views.test_autocomplete_view` doesn't pass due to a `DisallowedModelAdminToField` error being raised; I'm not sure how to resolve this issue as I'm not familiar with `to_field_allowed()` function.

File: tests/admin_widgets/test_autocomplete_widget.py [hunk 5]
```diff
@@ -53,7 +53,7 @@ def test_build_attrs(self):
             'class': 'my-class admin-autocomplete',
             'data-ajax--cache': 'true',
             'data-ajax--type': 'GET',
-            'data-ajax--url': '/admin_widgets/band/autocomplete/',
+            'data-ajax--url': '/admin_widgets/band/autocomplete/?to_field=id',
```
Comment (charettes) [inline tests/admin_widgets/test_autocomplete_widget.py]: I'd be great if `to_field` was not included if not necessary.

File: django/contrib/admin/widgets.py [hunk 6]
```diff
@@ -384,10 +384,12 @@ def __init__(self, rel, admin_site, attrs=None, choices=(), using=None):
         self.db = using
         self.choices = choices
         self.attrs = {} if attrs is None else attrs.copy()
+        self.to_field = self.rel.field.to_fields[0] if hasattr(self.rel.field, 'to_fields') else 'pk'
 
     def get_url(self):
         model = self.rel.model
-        return reverse(self.url_name % (self.admin_site.name, model._meta.app_label, model._meta.model_name))
+        url = reverse(self.url_name % (self.admin_site.name, model._meta.app_label, model._meta.model_name))
+        return url + '?to_field=' + self.to_field
```
Comment (charettes) [inline django/contrib/admin/widgets.py]: You can avoid adding `to_field` if not necessary by making the `self.to_field` assignment default to `None` if not present and only appending it if necessary here.

File: tests/admin_views/test_autocomplete_view.py [hunk 7]
```diff
@@ -25,22 +25,25 @@ class AuthorshipInline(admin.TabularInline):
     model = Authorship
     autocomplete_fields = ['author']
 
-
```
Comment (charettes) [inline tests/admin_views/test_autocomplete_view.py]: Revert unnecessary change.

File: tests/admin_views/test_autocomplete_view.py [hunk 8]
```diff
@@ -52,7 +55,7 @@ def setUpTestData(cls):
 
     def test_success(self):
         q = Question.objects.create(question='Is this a question?')
-        request = self.factory.get(self.url, {'term': 'is'})
+        request = self.factory.get(self.url, {'term': 'is'}) # return_id defaults to 'pk'
```
Comment (charettes) [inline tests/admin_views/test_autocomplete_view.py]: Not sure this is comment is required.

File: tests/admin_views/test_autocomplete_view.py [hunk 9]
```diff
@@ -61,6 +64,19 @@ def test_success(self):
             'results': [{'id': str(q.pk), 'text': q.question}],
             'pagination': {'more': False},
         })
+        # test for `to_field`
+        fk = ReferencedByParent.objects.create(name='ref by parent')
+        p = ParentWithFK.objects.create(fk=fk)
+        request = self.factory.get(self.url_to_field, {'term': 'by', 'to_field': 'fk'})
```
Comment (charettes) [inline tests/admin_views/test_autocomplete_view.py]: Nothing references `ParentWithFK.fk` so it's a disallowed field. An allowed field would be `ReferencedByParent.name` because it's referenced by `ParentWithFK.fk`.

File: django/contrib/admin/views/autocomplete.py [hunk 10]
```diff
@@ -24,12 +25,19 @@ def get(self, request, *args, **kwargs):
...
+        self.to_field = request.GET.get('to_field', None)
+        if self.to_field:
+            if not self.model_admin.to_field_allowed(request, self.to_field):
+                raise DisallowedModelAdminToField("The field %s cannot be referenced." % self.to_field)
+        else:
+            self.to_field = 'pk'
+
         self.paginator_class = self.model_admin.paginator
-        self.object_list = self.get_queryset()
+        self.object_list = self.get_queryset().order_by(self.to_field)
```
Comment (ConstantinoSchillebeeckx) [inline django/contrib/admin/views/autocomplete.py]: I'm ordering here because otherwise I get an `UnorderedObjectListWarning` in the `test_success` unit test
```
======================================================================
ERROR: test_success (admin_views.test_autocomplete_view.AutocompleteJsonViewTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/constantino/Desktop/django/tests/admin_views/test_autocomplete_view.py", line 73, in test_success
    response = AutocompleteJsonView.as_view(**self.as_view_args_to_field)(request)
  File "/Users/constantino/Desktop/django/django/views/generic/base.py", line 68, in view
    return self.dispatch(request, *args, **kwargs)
  File "/Users/constantino/Desktop/django/django/views/generic/base.py", line 88, in dispatch
    return handler(request, *args, **kwargs)
  File "/Users/constantino/Desktop/django/django/contrib/admin/views/autocomplete.py", line 38, in get
    context = self.get_context_data()
  File "/Users/constantino/Desktop/django/django/views/generic/list.py", line 119, in get_context_data
    paginator, page, queryset, is_paginated = self.paginate_queryset(queryset, page_size)
  File "/Users/constantino/Desktop/django/django/views/generic/list.py", line 58, in paginate_queryset
    allow_empty_first_page=self.get_allow_empty())
  File "/Users/constantino/Desktop/django/django/contrib/admin/views/autocomplete.py", line 49, in get_paginator
    return self.model_admin.get_paginator(self.request, *args, **kwargs)
  File "/Users/constantino/Desktop/django/django/contrib/admin/options.py", line 796, in get_paginator
    return self.paginator(queryset, per_page, orphans, allow_empty_first_page)
  File "/Users/constantino/Desktop/django/django/core/paginator.py", line 33, in __init__
    self._check_object_list_is_ordered()
  File "/Users/constantino/Desktop/django/django/core/paginator.py", line 125, in _check_object_list_is_ordered
    stacklevel=3
django.core.paginator.UnorderedObjectListWarning: Pagination may yield inconsistent results with an unordered object_list: <class 'admin_views.models.ReferencedByParent'> QuerySet.
```

Comment (charettes) [reply django/contrib/admin/views/autocomplete.py]: I think that's a different issue. I'd add a `Meta.ordering` to the test models to work around it for now.

File: tests/admin_widgets/test_autocomplete_widget.py (hunk 5 above)
Comment (ConstantinoSchillebeeckx) [reply tests/admin_widgets/test_autocomplete_widget.py]: I think it's necessary in this case because the FK field `band` is automatically assigned the `to_field='id'` since no `primary_key` is specified on `Band`.

**Comment by jpic** [comment]
Is this PR only blocked by isort and flake8 ?

**Comment by charettes** [comment]
@jpic that and the fact _patch needs improvement_ was never unchecked [on the ticket](https://code.djangoproject.com/ticket/29138).

**Comment by ConstantinoSchillebeeckx** [comment]
As far as I remember, I made all the requested changes and was just waiting on someone to give the 👍

**Comment by felixxm** [comment]
Superseded by #11990.
